filters/sed/editor.go

```diff
@@ -0,0 +1,474 @@
+package sed
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"regexp"
+)
+
+// TODO:
+// - we need to document that only non-zero matches are replaced
+// - we can document that it is currently not supported, together
+//   with the templating
+
+const (
+	readBufferSize             = 8 << 10
+	defaultMaxEditorBufferSize = 2 << 20
+)
+
+type maxBufferHandling int
+
+const (
+	maxBufferBestEffort maxBufferHandling = iota
+	maxBufferAbort
+)
+
+// editor provides a reader that wraps a input reader, and replaces each occurence of
+// the provided search pattern with the provided replacement. It can be used with a
+// delimiter or without.
+//
+// When using it with a delimiter, it reads enough data from the input until meeting
+// a delimiter or reaching maxBufferSize. The chunk includes the delimiter if any. Then
+// every occurence of the pattern is replaced, and the entire edited chunk is returned
+// to the caller.
+//
+// When not using a delimiter, it reads enough data until at least complete match of the
+// pattern is met or the maxBufferSize is reached. When the pattern matches the entire
+// buffered input, the replaced content is returned to the caller when maxBufferSize is
+// reached. This also means that more replacements can happen than if we edited the
+// entire content in one piece, but this is necessary to be able to use the editor for
+// input with unknown length.
+//
+// To limit the number of repeated scans over the buffered data, the size of the
+// additional data read from the input grows exponentially with every iteration that
+// didn't result with any edited data returned to the caller. If there was any edited
+// returned to the caller, the read size is reset to the initial value.
+//
+// When the input returns an error, e.g. EOF, the editor finishes editing the buffered
+// data, returns it to the caller, and returns the received error on every subsequent
+// read.
+//
+// When the editor is closed, it doesn't read anymore from the input or return any
+// buffered data. If the input implements io.Closer, closing the editor closes the
+// input, too.
+//
+type editor struct {
+	// init:
+	input             io.Reader
+	pattern           *regexp.Regexp
+	replacement       []byte
+	delimiter         []byte
+	maxBufferSize     int
+	maxBufferHandling maxBufferHandling
+	prefix            []byte
+	readBuffer        []byte
+
+	// state:
+	ready   *bytes.Buffer
+	pending *bytes.Buffer
+
+	// final:
+	err    error
+	closed bool
+}
+
+var (
+	ErrClosed           = errors.New("reader closed")
+	ErrEditorBufferFull = errors.New("editor buffer full")
+)
+
+func newEditor(
+	input io.Reader,
+	pattern *regexp.Regexp,
+	replacement []byte,
+	delimiter []byte,
+	maxBufferSize int,
+	mbh maxBufferHandling,
+) *editor {
+	if maxBufferSize <= 0 {
+		maxBufferSize = defaultMaxEditorBufferSize
+	}
+
+	rsize := readBufferSize
+	if maxBufferSize < rsize {
+		rsize = maxBufferSize
+	}
```

Comment:
Not sure. I don't really see the reason for that. `rsize` normally will be way smaller than `maxBufferSize`. They are different concerns. `rsize` controls how big chunks we try to read from the network at a time, and from experience 8k is a good value. `maxBufferSize` controls how much we allow to reside in the memory at a given time. The above safety check is only there because it never makes sense to read more than `maxBufferSize`. Does this make sense to you like this?

Reply:
fine for me, was just a suggestion.